File: maximum_difference_between_node_and_ancestor/c++/Binary_tree_inorder_traversal.cpp
//
//  Binary_tree_inorder_traversal.cpp
//  
//
//  Created by Wen-Ting Wang on 2018/11/7.
//

#include<iostream>
#include<stack>
#include<vector>
using namespace std;

/**
 * Definition for a binary tree node.
*/
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        if(!root) return ans;
        
        stack<TreeNode *> stackBag;
        TreeNode *current = root;
        
        while(current || !stackBag.empty()){
            
            while(current) { // search the left leaf
                stackBag.push(current);
                current = current->left;
            }
            current = stackBag.top();
            stackBag.pop();
            ans.push_back(current->val);
            
            current = current->right; // if left node is gone
        }
        return ans;
        
    }
};



//top-down
//ref http://www.voidcn.com/article/p-mquqvehe-bez.html
void Show(TreeNode *t) {
    if (!t) return;
    
    cout << t->val << " ";
    Show(t->left);
    Show(t->right);
}


int main(){
    TreeNode *nodeA = new TreeNode(1);
    TreeNode *nodeB = new TreeNode(2);
    TreeNode *nodeC = new TreeNode(3);
    TreeNode *nodeD = new TreeNode(4);
    TreeNode *nodeE = new TreeNode(5);
    TreeNode *nodeF = new TreeNode(6);
    
    nodeA->left = nodeB;
    nodeA->right = nodeC;
    nodeB->left = nodeD;
    nodeB->right = nodeE;
    nodeC->right = nodeF;
    
    Show(nodeA);
    vector<int> result, result2;
    Solution ans;
    
    result = ans.inorderTraversal(nodeA);
    
    
    cout<<endl;
    cout<<"Result1:"<<endl;
    for(int i = 0; i < 6; ++i)
        cout<<result[i]<<endl;
    
    
    
    return 0;
}

File: binaryTreeInorderTraversal/python/binary_tree_inorder_traversal.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    """
    Note
    ----
    - Time complexity: O(N)
    - Space complexity: O(N)
    """
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack_bag = []
        answer = []

        if root is None:
            return stack_bag

        while root or stack_bag:
            while root:
                stack_bag.append(root)
                root = root.left

            root = stack_bag.pop()
            answer.append(root.val)
            root = root.right

        return answer

class SolutionRecursive:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        return self.inorderTraversal(root.left) + [root.val] + self.inorderTraversal(root.right) if root else []

if __name__ == "__main__":
    tree = TreeNode(1)
    tree.right = TreeNode(2)
    tree.right.left = TreeNode(3)
    ans = Solution()
    print(f"The answer is {ans.inorderTraversal(tree)}")
    print(f"The recurisve answer is {SolutionRecursive().inorderTraversal(tree)}")

